### PR TITLE
Added managed server shutdown/startup in msi mode in Domain on PV model

### PR DESCRIPTION
**Pull request by anpanigr**
a. Added MSI Usecase (stop/start managed server) into ItIstioDomainOnPV class
b. Updated the ItServerStartPolicy to add the following usecase
       Managed server can not be started when spec.serverStartPolicy is ADMIN_ONLY or NEVER

Integration test results  
https://build.weblogick8s.org:8443/job/weblogic-kubernetes-operator-kind-new/3374/ 
https://build.weblogick8s.org:8443/job/weblogic-kubernetes-operator-kind-new/3378/ 
https://build.weblogick8s.org:8443/job/weblogic-kubernetes-operator-kind-new/3397/
https://build.weblogick8s.org:8443/job/weblogic-kubernetes-operator-kind-new/3404

### CODE REVIEW

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioDomainInPV.java [hunk 1]
```diff
@@ -135,7 +139,7 @@ public static void initAll(@Namespaces(2) List<String> namespaces) {
     labelMap.put("istio-injection", "enabled");
 
     assertDoesNotThrow(() -> addLabelsToNamespace(domainNamespace,labelMap));
-    assertDoesNotThrow(() -> addLabelsToNamespace(opNamespace,labelMap));
+    // assertDoesNotThrow(() -> addLabelsToNamespace(opNamespace,labelMap));
```
Comment (ankedia) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioDomainInPV.java:142]: Comment can be removed.

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioDomainInPV.java [hunk 2]
```diff
@@ -324,6 +332,49 @@ public void testIstioDomainHomeInPv() {
     boolean checkApp = checkAppUsingHostHeader(url, domainNamespace + ".org");
     assertTrue(checkApp, "Failed to access WebLogic application");
 
+    // Refer JIRA https://jira.oraclecorp.com/jira/browse/OWLS-86407
+    // Stop and Start the managed server in absense of administration server
+    assertTrue(patchServerStartPolicy(domainUid, domainNamespace,
+         "/spec/adminServer/serverStartPolicy", "NEVER"),
+         "Can't patch adminServer's serverStartPolicy to NEVER");
```
Comment (sankarpn) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioDomainInPV.java:339]:
```suggestion
         "Failed to patch adminServer's serverStartPolicy to NEVER");
```

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java [hunk 3]
```diff
@@ -245,7 +242,9 @@ public void testAdminServerRestart() {
     DateTime dynTs = getPodCreationTime(domainNamespace, dynamicServerPodName);
     DateTime cfgTs = getPodCreationTime(domainNamespace, configServerPodName);
 
-    patchServerStartPolicy("/spec/adminServer/serverStartPolicy", "NEVER");
+    assertTrue(patchServerStartPolicy(domainUid, domainNamespace,
+         "/spec/adminServer/serverStartPolicy", "NEVER"),
+         "Can't patch adminServer's serverStartPolicy to NEVER");
```
Comment (sankarpn) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java:247]:
```suggestion
         "Failed to patch adminServer's serverStartPolicy to NEVER");
```

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java [hunk 4]
```diff
@@ -271,7 +270,9 @@ public void testAdminServerRestart() {
     assertTrue(result.contains("script doesn't support starting or stopping administration server"),
         "The script shouldn't start the admin server");
 
-    patchServerStartPolicy("/spec/adminServer/serverStartPolicy", "IF_NEEDED");
+    assertTrue(patchServerStartPolicy(domainUid, domainNamespace,
+         "/spec/adminServer/serverStartPolicy", "IF_NEEDED"),
+         "Can't patch adminServer's serverStartPolicy to IF_NEEDED");
```
Comment (sankarpn) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java:275]:
```suggestion
         "Failed to patch adminServer's serverStartPolicy to IF_NEEDED");
```

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java [hunk 5]
```diff
@@ -427,9 +429,20 @@ public void testDomainRestart() {
...
+        "The script shouldn't start the managed server");
+    logger.info("Managed server instances can not be started while spec.serverStartPolicy is NEVER");
+    
     // Patch the Domain with serverStartPolicy set to ADMIN_ONLY
-    // Here only Admin server pod should come up
-    patchServerStartPolicy("/spec/serverStartPolicy", "ADMIN_ONLY");
+    // Here only Administration server pod should come up
+    assertTrue(patchServerStartPolicy(domainUid, domainNamespace,
+         "/spec/serverStartPolicy", "ADMIN_ONLY"),
+         "Can't patch domain's serverStartPolicy to ADMIN_ONLY");
```
Comment (sankarpn) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java:445]:
```suggestion
         "Failed to patch domain's serverStartPolicy to ADMIN_ONLY");
```

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java [hunk 6]
```diff
@@ -935,20 +978,20 @@ public void testDynamicServerLifeCycleWithoutAdmin() {
...
-   * MSI mode, the pre-requisite is the that the manged server MUST be started 
-   * once before admin server is shutdown, so that the embedded LDAP server is 
-   * replicated from admin server to the managed server. 
-   * In this case of MII and DomainInImage model, the server state/configuration
-   * is not saved once the server is shutdown unless we use domain-on-pv model
-   * So in MII case, startServer.sh script update the replica count but the 
-   * server startup is defered till we re-start the adminserver. Here the 
-   * operator tries to start the managed server but it will keep on failing 
-   * until AdminServer is available.   
+   * server from scrtach if it has never been started eralier with
```
Comment (sankarpn) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java:982]:
```suggestion
   * server from scratch if it has never been started earlier with
```

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java [hunk 7]
```diff
@@ -935,20 +978,20 @@ public void testDynamicServerLifeCycleWithoutAdmin() {
...
+   * administration Server. Once the administration server is stopped, the 
+   * managed server can only be started in MSI (managed server independence) 
+   * mode. To start a managed server in MSI mode, the pre-requisite is that the
+   * managed server MUST be started once before administration server is 
+   * shutdown, so that the security configuration is replicated to the managed 
+   * server. In this case of MII and DomainInImage model, the server 
+   * state/configuration is lost once the server is shutdown unless we use 
+   * domain-on-pv model. So in MII case, startServer.sh script update the 
+   * replica count but the server startup is defered till we re-start the
```
Comment (sankarpn) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java:991]:
```suggestion
   * replica count but the server startup is deferred till we re-start the 
```

**Review by sankarpn — CHANGES_REQUESTED** [review]

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java [hunk 8]
```diff
@@ -427,9 +429,20 @@ public void testDomainRestart() {
     result = executeLifecycleScript(STOP_DOMAIN_SCRIPT, DOMAIN, null);
     assertTrue(result.contains("No changes needed"), "stopDomain.sh shouldn't make changes");
 
+    // verify managed server instances can not be started while 
+    // spec.serverStartPolicy is NEVER
```
Comment (ankedia) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java:433]: I assume domain's spec.serverStartPolicy  is NEVER. Probably comment can say that "managed server instances can't be started while domain is stopped".

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java [hunk 9]
```diff
@@ -427,9 +429,20 @@ public void testDomainRestart() {
     result = executeLifecycleScript(STOP_DOMAIN_SCRIPT, DOMAIN, null);
     assertTrue(result.contains("No changes needed"), "stopDomain.sh shouldn't make changes");
```
Comment (ankedia) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java:430]: Use constant STOP_DOMAIN_SCRIPT for "stopDomain.sh"

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioDomainInPV.java [hunk 10]
```diff
@@ -255,6 +259,10 @@ public void testIstioDomainHomeInPv() {
                 .clusterName(clusterName)
                 .replicas(replicaCount)
                 .serverStartState("RUNNING"))
+            .addManagedServersItem(new ManagedServer()
+                .serverName("managed-server2")
```
Comment (ankedia) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioDomainInPV.java:263]: Replace with constant.

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioDomainInPV.java (hunk 2 above)
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioDomainInPV.java]: done

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java (hunk 3 above)
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java]: done

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java (hunk 4 above)
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java]: done

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java (hunk 5 above)
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java]: done

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java (hunk 6 above)
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java]: resolved the typo

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java (hunk 7 above)
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java]: resolved the typo

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java (hunk 9 above)
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java:430]: This is a string to be when the test  fails.

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioDomainInPV.java (hunk 1 above)
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioDomainInPV.java:142]: I should uncomment the statement. Istio should be enabled in Operator Namespace.

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java (hunk 8 above)
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java:433]: Modified the comments.

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioDomainInPV.java [hunk 11]
```diff
@@ -324,6 +332,49 @@ public void testIstioDomainHomeInPv() {
     boolean checkApp = checkAppUsingHostHeader(url, domainNamespace + ".org");
     assertTrue(checkApp, "Failed to access WebLogic application");
 
+    // Refer JIRA https://jira.oraclecorp.com/jira/browse/OWLS-86407
```
Comment (vanajamukkara) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioDomainInPV.java:329]: update the use case description for this change

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioDomainInPV.java [hunk 12]
```diff
@@ -324,6 +332,49 @@ public void testIstioDomainHomeInPv() {
...
+        domainUid, domainNamespace));
+    assertTrue(scalingSuccess,
+        String.format("Cluster scaling failed for domain %s in namespace %s", domainUid, domainNamespace));
+    logger.info("Cluster is scaled up in absense of administration server");
+    checkServiceExists(managedServerPodNamePrefix + "2", domainNamespace);
+    checkPodReady(managedServerPodNamePrefix + "2", domainUid, domainNamespace);
+    logger.info("Managed Server started in absense of administration server");
+  }
+
+  private void checkPodDeleted(String podName, String domainUid, String domNamespace) {
```
Comment (vanajamukkara) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioDomainInPV.java:360]: Check CommonTestUtils for this method and call that.

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java (hunk 8 above)
Comment (vanajamukkara) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java]: update the use case description for the changes

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioDomainInPV.java (hunk 10 above)
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioDomainInPV.java:263]: Managed Server need not be added, as I am doing scale down/up in the absence of Admin Server.

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java [hunk 13]
```diff
@@ -916,17 +956,20 @@ public void testDynamicServerLifeCycleWithoutAdmin() {
...
              adminServerPodName, domainUid, domainNamespace);
-      logger.info("AdminServer restart success");
+      logger.info("administration server restart success");
 
       checkPodReadyAndServiceExists(serverPodName, domainUid, domainNamespace);
       logger.info("(re)Started [" + serverName + "] on admin server restart");
     } finally {
       // restart admin server
-      patchServerStartPolicy("/spec/adminServer/serverStartPolicy", "IF_NEEDED");
+      assertTrue(patchServerStartPolicy(domainUid, domainNamespace,
```
Comment (vanajamukkara) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java:969]: you have the same code both inside try and in finally blocks, keep it at one place(finally).

**Review by sankarpn — APPROVED** [review]
LGTM

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java [hunk 14]
```diff
@@ -935,20 +978,20 @@ public void testDynamicServerLifeCycleWithoutAdmin() {
...
-   * MSI mode, the pre-requisite is the that the manged server MUST be started 
-   * once before admin server is shutdown, so that the embedded LDAP server is 
-   * replicated from admin server to the managed server. 
-   * In this case of MII and DomainInImage model, the server state/configuration
-   * is not saved once the server is shutdown unless we use domain-on-pv model
-   * So in MII case, startServer.sh script update the replica count but the 
-   * server startup is defered till we re-start the adminserver. Here the 
-   * operator tries to start the managed server but it will keep on failing 
-   * until AdminServer is available.   
+   * server from scratch if it has never been started eralier with
```
Comment (vanajamukkara) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java:981]:
```suggestion
   * server from scratch if it has never been started earlier with
```

Comment (vanajamukkara) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java:981]: what is the difference between testDynamicServerLifeCycleWithoutAdmin and testConfiguredServerLifeCycleWithoutAdmin? is it just the cluster type?

**Review by vanajamukkara — CHANGES_REQUESTED** [review]

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioDomainInPV.java (hunk 11 above)
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioDomainInPV.java:329]: Added the following line to the test description 

Additionally, the test verifies that WebLogic cluster can be scaled down and scaled up in the absence of Administration server.

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java (hunk 14 above)
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java:981]: yes. You are right

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java (hunk 13 above)
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java:969]: The code inside the try/catch block is a part of test, which verifies that the managed server get started once the admin server is (re) started. The finally block is for if the test fails for any other reason before reaching the above scenario.

Comment (vanajamukkara) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java:969]: finally block gets executed in all cases, you can just keep the code in finally. No need to keep the code in both try and finally.

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java (hunk 14 above)
Comment (vanajamukkara) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java]: Is there any reason for testing the same use case for both dynamic cluster and configured cluster? I don't think there is any difference in behavior based on cluster type.

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java (hunk 13 above)
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java:969]: I know. I want to start the stop/start the admin server in my test. The admin server start in finally block is void if the test pass.

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java (hunk 14 above)
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java]: Yes, we need coverage on both configured and dynamic cluster as far as server lifecycle is concerned.

**Review by ankedia — APPROVED** [review]
LGTM

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioDomainInPV.java (hunk 12 above)
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioDomainInPV.java:360]: updated tests to use CommonTestUtils .checkPodDeleted()

**Review by vanajamukkara — APPROVED** [review]

**Pull request merged**